### ISSUE TITLE
Texture scaling does not get applied on all textures

### ISSUE THREAD

**Issue by VIRGINKLM** [hrydgard/ppsspp#1934]
I had to put an issue on that weird bug. I have set texture filtering on XBRz x5 just to be sure it's not my imagination. I noticed on Sword Art Online - Infinity Moment, the characters on the speech sequences only their first texture gets filtered and the rest are left in nearest neighbor.
Take a look at this:

1st frame, texture gets properly filtered:
![ppssppwindows 2013-05-26 19-27-28-09](https://f.cloud.github.com/assets/3772761/565575/66f502a2-c622-11e2-96fc-c6aa0004e62f.png)

A couple of frames later from above frame, texture gets reset to nearest neighbor:
![ppssppwindows 2013-05-26 19-27-29-46](https://f.cloud.github.com/assets/3772761/565577/76b8c606-c622-11e2-8d21-b9d6903b9d5f.png)

**Comment by unknownbrackets** [comment]
Does this still happen?  I thought we made it regain trust with the max texels thing.

-[Unknown]

**Comment by daniel229** [comment]
Still happening,like when eyes blicking,the scaling effect will be gone.

**Comment by unknownbrackets** [comment]
Does it eventually come back, or is it gone permanently?

If the blink is animated, we probably can't reasonably scale the blink, especially if the texture is large...

-[Unknown]

**Comment by daniel229** [comment]
It won't come back.

**Comment by unknownbrackets** [comment]
If you remove these lines, it works right?  How is the performance?

```c++
				if (entry->numFrames < TEXCACHE_FRAME_CHANGE_FREQUENT) {
					entry->status |= TexCacheEntry::STATUS_CHANGE_FREQUENT;
				}
```

Maybe we should just rely entirely on texels per frame.

-[Unknown]

**Comment by daniel229** [comment]
Yes,it works.Some game may get much low performance,some games doesn't.
off
![off](https://cloud.githubusercontent.com/assets/3481559/6458083/516fa730-c1ba-11e4-8aa4-a49c1c76ab97.png)

2x
![2x](https://cloud.githubusercontent.com/assets/3481559/6458084/5171acd8-c1ba-11e4-942a-071d9bf91b51.png)

off
![off2](https://cloud.githubusercontent.com/assets/3481559/6458096/6559cc1c-c1ba-11e4-9b20-95c3697ff6e8.png)

2x
![2x2](https://cloud.githubusercontent.com/assets/3481559/6458095/6557070c-c1ba-11e4-97f3-c11b5669bee8.png)

**Comment by unknownbrackets** [comment]
What if, instead, you decrease this number (in TextureCache.h) to something more like 60?

const static int FRAMES_REGAIN_TRUST = 1000;

-[Unknown]

**Comment by daniel229** [comment]
60 won't work,it work a moment when blicking then disappear.

**Comment by Kolrath** [comment]
I've notcied this occuring in numerous 2D games when using filtering.  I had always chalked it up as being a glitch related to PPSSPP as I've not noticed it to occur in 2D games on other system emulators such as the PS2 emulator PCSX2.

From what I'm gathering from reading in this thread, as well as issue #2840 and #4235 mentioned above, is that this is the result of PPSSPP not filtering things that are regularly being updated as it would impact performance.  Is this understanding correct?

If I am understanding that correctly then I've got a further question of... what is going on with the game Astonishia Story in relation to filtering?  

When using XBRz, regardless of whether its 2x, 5x, auto, or whatever, it seems that only the animated objects are having the scaling applied while any static objects are left being pixelated.

In the example image below the two character sprites are being filtered as is the clock and the two barrels with water in them.  These are all objects that are animated and constantly changing.  However, the static objects like.... well everything else actually... are left unfiltered.

Would I be right in assuming this isn't the intended function?

![astonishia story filtering example](https://cloud.githubusercontent.com/assets/8617529/10179405/beafe1c0-66d3-11e5-9153-e73a4a8c645c.jpg)

I tried messing around with countless combinations of the various filtering and scaling options to see if anything had any different results but in all cases the animated stuff was being filtered and the static stuff was not.

I'm noticed sporadic / odd filtering happening in numerous other 2D games and from the standpoint of casual observation it seems fairly random.  For example, in the game Grownlanser IV - Wayfarer of time some objects and characters would be filtered and others wouldn't.  It seemed random in that sometimes the players moving character would or wouldn't be filtered or sometimes just a couple static objects or characters wouldn't be filtered or when two people are attacking each other and both are animated only one of them would be being filtered, etc.

**Comment by LunaMoo** [comment]
^ You mix filtering with scaling, but I guess you mean the latter one all the way. PPSSPP has limits on texture scaling for performance, both in frequency the texture changes and on amount of work that it can do per frame. Without them, some games can easily fill a few gb's of video ram and put constant abuse on cpu, in other words lag like crazy even on very powerful modern desktops.

 However your example doesn't really show anything wrong with how those limits work ~ each game is different and especially 2D games are often working in very weird ways.
 This game looking at the screenshot for example could be using cpu to generate background graphics as a single texture starting from small tiles and that resulting texture would be slightly different every frame or two either by some effect or even by some trash data from game memory it copies the texture to/from. Even unnoticeable difference matters as it really must be pixel perfect to be threated as same texture and so not ignored by texture scaling.
 Animated objects in this one more than likely keep all frames of animation as one simple texture per object or maybe even whole page of them that never changes that's why scaling works on it.

 You can just use GE debugger(on windows at least) ctrl+G or inside debug menu to observe step by step how graphics is made and see for yourself how the background is made.
 The function certainly does work as intended, the result in some games might be far from ideal through, that's why an issue like this exists in the first place.

**Comment by unknownbrackets** [comment]
They may seem static, but they may actually not be.  For example, many 2D games:

A. Have animated sprites, e.g. images of people or animals that move around.
A1. They may change these images for each animation frame.  IIRC, Tales of Eternia does this.
A2. They may keep separate frames all available as separate textures, allowing animation to use scaling and long-term cache.

B. Have background graphics, as a tilesheet.
B1. This may be entirely static, which should be cacheable.
B2. This may only draw static parts of the texture, in an "L" shape.  Other areas of the texture may get updated for temporary usage, just because it's "free" memory to squirrel away data into.  This counts as a change to the texture, even if that changed part is never rendered.
B3. This may be animated.  Parts of the tilesheet may be updated to account for e.g. rippling water or fireplaces.  Whether drawn or not drawn, these count as changes to the texture.  Popolocrois does this, for example.

So just because it uses a friendly method for the sprites doesn't mean it does for the tiles.  Some of the above is just as @LunaMoo mentioned, I'm just clarifying why it's not necessary that both parts will behave the same way.

-[Unknown]

**Comment by Kolrath** [comment]
Thank you both for the information.  All this kinda stuff is still a good bit "alien" to me so I don't really fully grasp how everything works.  Not overly surprised that I was mixing up my usage of filtering and scaling.

This is likely another stupid question here, but I'm curious...

Instead of address each individual sprite / texture, would it be feasible to achieve similar result by just taking the entire image / surface before its rendered to the screen and modifying it?  I was always under the impression that post-processing effects were less demanding since they only need to be done once to the "finished" image before its displayed on screen.  I'm aware that wouldn't exactly be the same thing as the currently filtering / scaling methods, but it seems like it could result in the kinds of image quality that people are expecting to see based on their example screenshots in this thread.

Sorry if its a dumb question.  Just curious about it.

**Comment by LunaMoo** [comment]
Yeah well we have a separate issue opened for that #6280
 There are some postprocessing shader versions of xBR and other upscaling shaders which for fully 2D games could possibly work better than texture scaling, the thing is nobody made any good one for PPSSPP as of yet.

 Some other emulators do have it and I belive libretro as well which do have ppsspp core(?) so maybe you could use it to get a functional full screen upscaling postprocessing shader on ppsspp. Never tried using libretro through, so not really sure about that. I did tried to port some existing xBR postprocessing shaders for ppsspp in the past, but maybe I'm just too stupid to understand how the scaling should happen through postprocessing as I was only able to get xBR rounding still in low res.;p

 PPSSPP does have one upscaling shader ~ Spline36 upscaler, you can try it - if I'm not messing that up technically it should upscale when you use low rendering res(x1) and high window size/fullscreen, in practice, well just try it.
 As for me it kind of just shifts pixels which I guess would be a simple sharpen method if the shift was more subtle and probably it would be if it was done after scaling, so I wonder if that shader doesn't have same problem as my failed tries to make upscaling shader and does not scale at all heh. Then again Henrik made that one, so it might work fine and just look ugly in psp games.

**Comment by hrydgard** [comment]
I didn't write Spline36, it was contributed. Either way it would indeed be nice with more post shaders. It's possible there are some issues with how scaling works, but it should be possible to get some xBR type shader working. If not there's a bug.

**Comment by LunaMoo** [comment]
Oh sorry, maybe I just remembered you doing some changes to it when I first saw it:].

 I tried again today, this time I found 5xBR v3.5a already in GLSL, so there's even less to do(and potentially break) than porting from .cg/hlsl which I tried in the past, yet I end up with black screen and can't find the reason for it, there's no error, I'm setting alpha to 1.0 as well heh, maybe I just messed up coordinates somehow and it get's off screen.

I used version posted by "guest"(that's his actual nickname, not some random person;p) near the bottom in http://www.razyboard.com/system/morethread-xbr-shader-pete_bernert-266904-6143523-0.html

pretty much the difference from PPSSPP would be those OGL2Size and OGL2InvSize, which I found mean:
OGL2InvSize.xy: 1/texture_size.xy
OGL2Size.x = texture x size (for example "2048" in "high internal resolution" setting)
OGL2Size.y = texture y size (for example "1024" in "high internal resolution" setting)

which I belive would in PPSSPP translate to:
OGL2InvSize.xy = u_pixelDelta.xy;
OGL2Size.xy = 1.0 / u_pixelDelta.xy;
u_pixelDelta is pretty much equal to pixel size on the screen with OutputResolution=True, or I messed that up?

**Comment by daniel229** [comment]
Comment out this line not slow down the game now,but slow down the video,stop apply to video then may have a try.
```
                if (entry->numFrames < TEXCACHE_FRAME_CHANGE_FREQUENT) {
                    entry->status |= TexCacheEntry::STATUS_CHANGE_FREQUENT;
                }
```

![01](https://cloud.githubusercontent.com/assets/3481559/12006127/af6554dc-ac03-11e5-9e75-e0188285547f.png)
![02](https://cloud.githubusercontent.com/assets/3481559/12006128/af70e7ac-ac03-11e5-9099-25eaf953f551.png)

**Comment by daniel229** [comment]
Some scene may still slow much,and still some scenes not get the effect.

**Comment by unknownbrackets** [comment]
How is it if you increase `TEXCACHE_FRAME_CHANGE_FREQUENT` from 6 to maybe 18?

-[Unknown]

**Comment by daniel229** [comment]
increase TEXCACHE_FRAME_CHANGE_FREQUENT from 6 to 18 didn't work.

**Comment by unknownbrackets** [comment]
Is it any better in the latest git build?  Note that some textures may not *immediately* scale, mainly due to performance concerns.  But I made it so they eventually scale much sooner, if possible.

@LunaMoo if you're still working on that shader, we could probably use it in two areas:

1. When games upload RAM to framebuffers via block transfers, we could scale using that shader.
2. As a post-shader.

The first option might be interesting for cases where the game downloads, and then re-uploads an area, to do effects on it.  It might work better in games where the post-shader is not always desirable.  Just an idea.

-[Unknown]

**Comment by LunaMoo** [comment]
Well 5xBR is already included in PPSSPP as post shader: https://github.com/hrydgard/ppsspp/blob/master/assets/shaders/5xBR.fsh
 Downsides of it are kind of obvious - certainly low res for 3D stuff and it also doesn't look as good as texture scaling with "pixel art" or any other stretched low res textures since using rendering res for pixel size.

 That leaves 1 - would certainly not hurt to have an option like that at least some games could benefit from it, I have no idea how to do that through.

**Comment by unknownbrackets** [comment]
Oh, I thought xBR was different from xBRZ and misread your comment as being about xBRZ.

For 1 - that's DrawPixels.  Every case that DrawPixels is called now is when rendering RAM into a framebuffer or the screen.  Right now it always uses `draw2dprogram_` for that.  So the change would be making it use a different program, one probably initialized at the same time as the postshader (but probably a separate setting from the postshader.)

-[Unknown]

**Comment by unknownbrackets** [comment]
How is this looking now?  We could implement something to skip scaling videos, although I worry that doesn't solve the general problem...

To some extent, texture replacement can now be used to work around this problem.  Not calling that a fix exactly, but pre-scaled texture loading will perform better so doesn't need the same limits.

-[Unknown]

**Comment by daniel229** [comment]
The effect goes away as soon as it blinks,then comes back after 0.5 second.

**Comment by Q3CPMA** [comment]
Hello, is this normal if this is still happening with 1.2.2?

**Comment by unknownbrackets** [comment]
Yes, old releases are not retroactively modified.  1.2.2 isn't going to get changes made after it was released.

Also, this issue is still marked open, so it would of course be normal to still see it.

In either case, texture scaling is very expensive (especially at high scaling levels), and some games change textures every frame.  This issue is really about scaling as much as is realistically possible - so if you're waiting for texture scaling in a game like FF2, you might need to get a 4.5Ghz 48-core CPU first.

-[Unknown]

**Comment by Q3CPMA** [comment]
Nah, I came here from the Disgaea (1 & 2) issue.

**Comment by ResurrectionRedmi9** [comment]
Auto Max Quality in texture scaling could help this issue?

**Comment by hrydgard** [comment]
This has been fixed, closing.
